cpp/arcticdb/util/test/test_storage_lock.cpp

```diff
@@ -182,6 +182,46 @@ struct ForceReleaseLockTask {
     }
 };
 
+struct OptimisticForceReleaseLockTask {
+    std::shared_ptr<LockData> data_;
+    size_t timeout_ms_;
+    size_t retry_ms_;
+
+    OptimisticForceReleaseLockTask(std::shared_ptr<LockData> data, size_t timeout_ms, size_t retry_ms) :
+            data_(std::move(data)),
+            timeout_ms_(timeout_ms),
+            retry_ms_(retry_ms)
+    {
+    }
+
+    folly::Future<folly::Unit> operator()() {
+        std::this_thread::sleep_for(std::chrono::milliseconds(1));
+        StorageLock<> lock{data_->lock_name_};
+        bool contended = true;
+        size_t total_wait = 0;
+        while (contended && total_wait < timeout_ms_) {
+            contended = !lock.try_lock(data_->store_);
+            std::this_thread::sleep_for(std::chrono::milliseconds(retry_ms_));
+            total_wait += retry_ms_;
+
+        }
+        if (contended) {
+            data_->timedout_ = true;
+            data_->contended_ = true;
+        } else {
+            ++data_->vol_;
```

Comment:
Volatile does not protect from data races, moreover volatile variables are not atomic and must be protected by a lock when concurrent reads and writes are done. The only thing that volatile does is to prevent the compiler from optimizing the value away it has no relation to the (concurrent) memory model. In the test the check data_->vol_ == 4 will produce a flaky test.

Reply:
The idea is to have a volatile member that we *want* to be corrupted/not incremented correctly in the case that the storage lock is not doing its job properly.

The atomic one is then a comparison, since this should definitely be updated correctly due to the nature of atomic type in C++.

I agree we don't need to wrap the atomic in a mutex, that is just how it was always done, and I'm not sure why. I think it's a mistake. If you agree should I remove this mutex? I.e. all uses of `data_->mutex_`?

Reply:
> The idea is to have a volatile member that we want to be corrupted/not incremented correctly in the case that the storage lock is not doing its job properly.

I'm not following on this. In the test you're doing the same check for the vol_ as for atmoic_. If the lock is not doing the correct thing vol_ will be a random value between 1 and 4

>  If you agree should I remove this mutex?
Remove it in all places where it guards the atomic, just make sure it's not guarding anything else.